File: Java/src/cn/z/PinyinUtil.java
package cn.z;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

public class PinyinUtil {
	private static final Map<Character, List<Integer>> charMap;
	private static final Map<Integer, String> pinyinMap;

	static {
		charMap = new HashMap<>();
		try {
			Properties charProp = new Properties();
			charProp.load(new FileInputStream("data/pinyinmap.properties"));
			Enumeration<?> e = charProp.propertyNames();
			while (e.hasMoreElements()) {
				String key = (String) e.nextElement();
				String indexStr = charProp.getProperty(key);
				String[] indexArr = indexStr.split(",");
				List<Integer> indexs = new ArrayList<>();
				for (String index : indexArr) {
					indexs.add(Integer.parseInt(index));
				}
				charMap.put(key.charAt(0), indexs);
			}
		} catch (IOException e) {
			e.printStackTrace();
		}

		pinyinMap = new HashMap<>();

		try {
			Properties indexProp = new Properties();
			indexProp.load(new FileInputStream("data/pinyinindex.properties"));
			Enumeration<?> e = indexProp.propertyNames();

			while (e.hasMoreElements()) {
				String key = (String) e.nextElement();
				String indexStr = indexProp.getProperty(key);
				pinyinMap.put(Integer.parseInt(key), indexStr);
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	/**
	 * 
	 * @param c
	 * @return 最常用的一个拼音，小写
	 */
	public static String toHanyupinyin(char c) {
		List<Integer> indexs = charMap.get(c);
		if (indexs != null) {
			String py = pinyinMap.get(indexs.get(0));
			return py.substring(0, py.length() - 1);
		}
		return String.valueOf(c);
	}
	
	/**
	 * 
	 * @param str
	 * @return 汉字转拼音
	 */
	public static String toHanyupinyin(String str) {
	    if (str == null) {
	    	return "";
	    }
	    StringBuilder sb = new StringBuilder();
	    char[] arr = str.toCharArray();
	    boolean isFirst = true;
		for(char c : arr) {
			if (isFirst){
				isFirst = false;
			}
			else {
				sb.append(" ");
			}
			sb.append(toHanyupinyin(c));
		}
		return sb.toString();
	}

	/**
	 * 
	 * @param c
	 * @return 全部拼音，带声调
	 */
	public static List<String> getAllPinyin(char c) {
		List<Integer> indexs = charMap.get(c);
		List<String> pys = new ArrayList<>();
		if (indexs != null) {
			for (Integer index : indexs) {
				String py = pinyinMap.get(index);
				pys.add(py);
			}
			return pys;
		}
		return pys;
	}

	private static List<Integer> match(final String str, final String key) {
		System.out.println("matching: " + str + " " + key);
		if (key.length() == 0) {
			return new ArrayList<>();
		}
		List<Integer> matchAllIndex = new ArrayList<>();
		List<Integer> matchIndex = new ArrayList<>();
		int j = 0;
		for (int i = 0; i < str.length() && j < key.length(); i++) {
			if (key.charAt(j) == str.charAt(i)) {
				matchIndex.add(i);
				j++;
			}
			if (j == key.length()) {
				matchAllIndex.addAll(matchIndex);
				matchIndex = new ArrayList<>();
				j = 0;
			}
		}

		return matchAllIndex;
	}

	/**
	 * 
	 * @param str
	 * @param searchKey
	 * @return 搜索匹配的index，可以用于高亮显示结果
	 */
	public static List<Integer> matchOrPinyinMatch(String str, String searchKey) {
		String name = str.toLowerCase();
		String key = searchKey.toLowerCase();
		List<Integer> ret = match(name, key);
		if (ret.size() > 0) {
			return ret;
		}
		List<Integer> pyIndex = new ArrayList<>();
		int startIndex = 0;
		StringBuilder pyStr = new StringBuilder();
		for (int i = 0; i < name.length(); i++) {
			String py = toHanyupinyin(name.charAt(i));
			pyStr.append(py);
			for (int j = 0; j < py.length(); j++) {
				pyIndex.add(startIndex);
			}
			startIndex++;
		}

		ret = match(pyStr.toString(), key);
		if (ret.size() > 0) {
			List<Integer> pyRet = new ArrayList<>();
			for (Integer index : ret) {
				Integer oriIndex = pyIndex.get(index);
				if (!pyRet.contains(oriIndex)) {
					pyRet.add(oriIndex);
				}
			}

			return pyRet;
		}

		return new ArrayList<>();
	}
	
	public static String hilightSearch(String str, String searchKey) {
		StringBuilder sb = new StringBuilder();
		List<Integer> m = matchOrPinyinMatch(str,searchKey);
		for(int i = 0; i < str.length(); i++) {
			if (m.contains(i)) {
				sb.append("<em>").append(str.charAt(i)).append("</em>");
			}
			else {
				sb.append(str.charAt(i));
			}
		}
        return sb.toString();
    }

}

File: Kotlin/src/cn/z/PinyinUtil.kt
package cn.z

import java.io.File
import java.util.Properties
import java.io.FileInputStream

class PinyinUtil {
	companion object {
		val charMap: MutableMap<Char, List<Int>> = HashMap();
		val pinyinMap: MutableMap<Int, String> = HashMap();

		init {
			//here goes static initializer code

			var charProp = Properties()
			charProp.load(FileInputStream("data/pinyinmap.properties"));
			var e = charProp.propertyNames();
			while (e.hasMoreElements()) {
				var key = e.nextElement() as String;
				var indexStr = charProp.getProperty(key);
				var indexs = indexStr.split(",").map { index -> Integer.parseInt(index) };
				charMap.put(key.elementAt(0), indexs);
			}

			charProp = Properties()
			charProp.load(FileInputStream("data/pinyinindex.properties"));
			e = charProp.propertyNames();
			while (e.hasMoreElements()) {
				var key = e.nextElement() as String;
				var indexStr = charProp.getProperty(key);

				pinyinMap.put(Integer.parseInt(key), indexStr);
			}
		}

		/**
		 * 最常用的一个拼音，小写
		 */
		fun toHanyupinyin(c: Char): String {
			var indexs = charMap.get(c);
			if (indexs != null) {
				var py = pinyinMap.get(indexs.get(0));
				return py!!.substring(0, py.length - 1);
			}
			return c.toString();
		}

		/**
		 * 全部拼音，带声调
		 */
		fun getAllPinyin(c: Char): List<String> {
			val indexs = charMap.get(c);
			val pys = ArrayList<String>();
			if (indexs != null) {
				for (index in indexs) {
					val py = pinyinMap.get(index);
					pys.add(py!!);
				}
			}
			return pys;
		}

		private fun match(str: String, key: String): List<Int> {
			println("matching: " + str + " " + key)
			if (key.length == 0)
				ArrayList<Int>()
			val matchIndex = ArrayList<Int>()
			var j = 0
			var i = 0
			while (i < str.length && j < key.length) {
				if (key.get(j) == str.get(i)) {
					matchIndex.add(i)
					j++
				}
				i++
			}
			if (j == key.length) {
				return matchIndex
			}
			return ArrayList<Int>()
		}

		/**
		 *
		 * @param str
		 * @param searchKey
		 * @return 搜索匹配的index，可以用于高亮显示结果
		 */
		fun matchOrPinyinMatch(str: String, searchKey: String): List<Int> {
			val name = str.toLowerCase()
			val key = searchKey.toLowerCase()
			var ret = match(name, key)
			if (ret.size > 0) {
				return ret
			}
			val pyIndex = ArrayList<Int>()
			var startIndex = 0
			val pyStr = StringBuilder()
			for (i in 0..name.length - 1) {
				val py = toHanyupinyin(name.get(i))
				pyStr.append(py)
				for (j in 0..py.length - 1) {
					pyIndex.add(startIndex)
				}
				startIndex++
			}
			ret = match(pyStr.toString(), key)
			if (ret.size > 0) {
				val pyRet = ArrayList<Int>()
				for (index in ret) {
					val oriIndex = pyIndex.get(index.toInt())
					if (!pyRet.contains(oriIndex)) {
						pyRet.add(oriIndex)
					}
				}
				return pyRet
			}
			return ArrayList<Int>()
		}
	}
}